datalad/local/remove.py

```diff
@@ -0,0 +1,229 @@
+            # we must not offer a 'nothing' which would bypass
+            # the `drop()` call. The implementation completely
+            # relies on `drop()` for all safety measures.
+            # instead `drop(reckless=kill)` must be used to fast-kill
+            # things
+            constraints=EnsureChoice('datasets', 'all')),
+        jobs=jobs_opt,
+        message=save_message_opt,
+        # XXX deprecate!
+        save=Parameter(
+            args=("--nosave",),
+            dest='save',
+            action="store_false",
+            doc="""DEPRECATED and IGNORED; use `save --amend` instead"""),
+        recursive=Parameter(
+            args=("--recursive", '-r',),
+            doc="""DEPRECATED and IGNORED: removal is always a recursive
+            operation"""),
+    )
+    # inherit some from Drop
+    # if_dirty and check as deprecated
+    for p in ('reckless', 'if_dirty', 'check',):
+        _params_[p] = Drop._params_[p]
+
+    _examples_ = [
+        dict(text="Permanently remove a subdataset from a dataset",
+             code_py="remove(dataset='path/to/dataset', path='path/to/subds')",
+             code_cmd="datalad remove -d <path/to/dataset> <path/to/subds>"),
+    ]
+
+    @staticmethod
+    @datasetmethod(name='remove')
+    @eval_results
+    def __call__(
+            path=None,
+            dataset=None,
+            drop='datasets',
+            reckless=None,
+            message=None,
+            jobs=None,
+            # deprecated below
+            recursive=None,
+            check=None,
+            save=None,
+            if_dirty='save-before'):
+
+        # deprecate checks
+        if check is False:
+            reckless = 'availability'
+
+        refds = require_dataset(dataset, check_installed=True,
+                                purpose='remove')
+        # same path resolution that drop will do
+        paths_by_ds, errors = get_paths_by_ds(
+            refds, dataset, ensure_list(path),
+            # super-mode will readily tell us which datasets to
+            # save as the end
+            subdsroot_mode='super')
+
+        drop_success = True
+        for res in Drop.__call__(
+                dataset=dataset,
+                path=path,
+                what=drop,
+                reckless=reckless,
+                recursive=True,
+                recursion_limit=None,
+                jobs=jobs,
+                result_xfm=None,
+                return_type='generator',
+                result_renderer='disabled',
+                # delegate error handling here
+                on_failure='ignore'):
+            if res.get('status') not in ('ok', 'notneeded'):
+                drop_success = False
+            yield res
+
+        if not drop_success:
+            # there will be 'rm -rf' below, so play safe
+            lgr.debug('Observed drop failure, will not attempt remove')
+            return
+
+        for dpath, paths in paths_by_ds.items():
```

Comment:
Not yet sure where this problem is rooted, but ``remove`` fails to save a subdataset deletion if it is a hierarchy if datasets:

```
# remove a single level of subdatasets is fine:
❱ datalad get -n HCP1200/996782   
[INFO   ] Fetching 'http://store.datalad.org/89d/4899e-1fe6-11ea-9e62-002590496000/config' 
install(ok): /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782 (dataset) [Installed subdataset in order to get /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782]
❱ datalad remove HCP1200/996782                                                                                                   
uninstall(ok): HCP1200/996782 (dataset)
remove(ok): HCP1200/996782 (file)                                                                                                     
save(ok): . (dataset)                                                                                                                 
action summary:                                                                                                                       
  remove (ok: 1)
  save (ok: 1)
  uninstall (ok: 1)
```


```
# removal of many levels of subdatasets leaves the superdataset dirty:
❱ datalad get -n -r HCP1200/996782

[INFO   ] Installing Dataset(/home/adina/repos/human-connectome-project-openaccess/HCP1200/996782) to get /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782 recursively 
[INFO   ] scanning for annexed files (this may take some time) 
[INFO   ] Fetching 'http://store.datalad.org/937/b2e76-1fe6-11ea-8fe3-002590496000/config' 
install(ok): /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782/MNINonLinear (dataset)                              
[INFO   ] Fetching 'http://store.datalad.org/94f/c881c-1fe6-11ea-8fe3-002590496000/config'                                            
install(ok): /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782/T1w (dataset)                                       
[INFO   ] Fetching 'http://store.datalad.org/96a/2d630-1fe6-11ea-8fe3-002590496000/config'                                            
install(ok): /home/adina/repos/human-connectome-project-openaccess/HCP1200/996782/unprocessed (dataset)                               
action summary:                                                                                                                       
  get (notneeded: 1)
  install (ok: 3)
datalad get -n -r HCP1200/996782  5.83s user 3.39s system 32% cpu 27.943 total
(handbook2) adina@muninn in ~/repos/human-connectome-project-openaccess on git:master
❱ datalad remove -d HCP1200/996782
uninstall(ok): MNINonLinear (dataset)
uninstall(ok): T1w (dataset)
uninstall(ok): unprocessed (dataset)
uninstall(ok): . (dataset)
action summary:
  uninstall (ok: 4)
(handbook2) adina@muninn in ~/repos/human-connectome-project-openaccess on git:master!
❱ git st
On branch master
Your branch is ahead of 'origin/master' by 1 commit.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	deleted:    HCP1200/996782
```

Reply:
> `datalad remove -d HCP1200/996782`

You are not removing *from* the superdataset, hence it stays unsaved. Without `-d`:

```sh
(datalad-dev) mih@meiner /tmp/hcp (git)-[master] % datalad remove HCP1200/996782
uninstall(ok): HCP1200/996782/MNINonLinear (dataset)
uninstall(ok): HCP1200/996782/T1w (dataset)
uninstall(ok): HCP1200/996782/unprocessed (dataset)
uninstall(ok): HCP1200/996782 (dataset)
remove(ok): HCP1200/996782 (file)                                                                                           
save(ok): . (dataset)                                                                                                       
action summary:                                                                                                             
  remove (ok: 1)
  save (ok: 1)
  uninstall (ok: 4)
(datalad-dev) mih@meiner /tmp/hcp (git)-[master] % git st
On branch master
Your branch is ahead of 'origin/master' by 1 commit.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
```

Reply:
ehhh right :facepalm: